bootstrap/updater.go

```diff
@@ -154,6 +155,8 @@ func (u *Updater) Start(ctx context.Context) error {
 		return fmt.Errorf("data dir not set %s", u.cfg.DataDir)
 	}
 	u.once.Do(func() {
+		ctx, cancel := context.WithCancel(ctx)
+		u.stop = cancel
```

Comment:
I don't think this is safe, Start and Close are called in different goroutines and so there's no guarantee that Close will see any change to u.stop, and also u.stop may not have been set by by the time that Close calls it.

Reply:
The correct fix doesn't seem like such a big change to me.

* Pass the context at construction time instead of in Start.
* Make close just call the cancel function of the context.
* Cleanup channels when exiting Start.

Seems like it should work

Reply:
> Pass the context at construction time

No. This is not how contexts are intended be used. If a function/method returns it should not use the passed in context at all any more.

Reply:
See also: https://go.dev/blog/context-and-structs

If the function/method that receives the context returns it should not hold on to a reference of the context any more, also not onto child contexts. I just tried to limit the scope of my change but `Start` and `Close` shouldn't use any context at all.

Reply:
Thanks for the link, it seems that we are misusing contexts throughout the codebase!

Reply:
This is indeed the case but this PR is already quite large and I prefer multiple small PRs over a few large 🙂

Reply:
That being said I addressed this immediate concern by removing the passed in context from `Start` and use a channel to indicate closure over a `context.CancelFunc`.

Reply:
Yes .. but! The purpose of this PR is to fix the unit tests and this change introduces a new problem that could cause tests to fail randomly. So I'd really be inclined to fix that as part of the PR. I not going to block it on this but lets see what @poszu thinks?

Reply:
As I said, I addressed this immediate concern, but I won't update the whole codebase to use context correctly 😉

Reply:
Oh sorry! Great, I misread your comment :D